src/Streamly/Internal/Data/Fold/Types.hs

```diff
@@ -64,7 +90,26 @@ import Streamly.Internal.Data.SVar (MonadAsync)
 
 data Fold m a b =
   -- | @Fold @ @ step @ @ initial @ @ extract@
-  forall s. Fold (s -> a -> m s) (m s) (s -> m b)
+  forall s. Fold (s -> a -> m (Step s b)) (m s) (s -> m b)
+
+{-# INLINE stepWS #-}
+stepWS :: Monad m => (s -> a -> m (Step s b)) -> Step s b -> a -> m (Step s b)
+stepWS step (Yield s) a = step s a
```

Comment:
stepWS/doneWS/initialTS/initialTSM sound pretty cryptic. Can we use alternate names to have a consistent lifting/wrapping concept e.g.:

```
-- lift the first argument of `step` from `s` to `Step s b`
liftStep :: ...
-- lift the first argument of `extract` from `s` to `Step s b`
liftExtract :: ...
-- lift `s` to `Step s b`
liftInitial :: ...
```